framework/src/systems/LinearSystem.C

```diff
@@ -0,0 +1,293 @@
+#include "libmesh/petsc_matrix.h"
+#include "libmesh/default_coupling.h"
+#include "libmesh/diagonal_matrix.h"
+
+#include <ios>
+
+namespace Moose
+{
+void
+compute_linear_system(libMesh::EquationSystems & es, const std::string & system_name)
+{
+  FEProblemBase * p = es.parameters.get<FEProblemBase *>("_fe_problem_base");
+  auto & sys = p->getLinearSystem(p->linearSysNum(system_name));
+  auto & lin_sys = sys.linearImplicitSystem();
+  auto & matrix = *(sys.linearImplicitSystem().matrix);
+  auto & rhs = *(sys.linearImplicitSystem().rhs);
+  p->computeLinearSystemSys(lin_sys, matrix, rhs);
+}
+}
+
+LinearSystem::LinearSystem(FEProblemBase & fe_problem, const std::string & name)
+  : SolverSystem(fe_problem, fe_problem, name, Moose::VAR_SOLVER),
+    PerfGraphInterface(fe_problem.getMooseApp().perfGraph(), "LinearSystem"),
+    _sys(fe_problem.es().add_system<LinearImplicitSystem>(name)),
+    _rhs_time_tag(-1),
+    _rhs_time(NULL),
+    _rhs_non_time_tag(-1),
+    _rhs_non_time(NULL),
+    _n_linear_iters(0),
+    _linear_implicit_system(fe_problem.es().get_system<LinearImplicitSystem>(name)),
+    _solution_is_invalid(false)
+{
+  getRightHandSideNonTimeVector();
+  // Don't need to add the matrix - it already exists (for now)
+  _system_matrix_tag = _fe_problem.addMatrixTag("SYSTEM");
+
+  // We create a tag for the right hand side, the vector is already in the libmesh system
+  _rhs_tag = _fe_problem.addVectorTag("RHS");
+
+  _linear_implicit_system.attach_assemble_function(Moose::compute_linear_system);
+}
+
+LinearSystem::~LinearSystem() = default;
+
+void
+LinearSystem::addTimeIntegrator(const std::string & /*type*/,
+                                const std::string & /*name*/,
+                                InputParameters & /*parameters*/)
+{
+  mooseError("LinearSystem does not support time integrators yet!");
+}
+
+void
+LinearSystem::initialSetup()
+{
+  SystemBase::initialSetup();
+  // Checking if somebody accidentally assigned nonlinear variables to this system
+  const auto & var_names = _vars[0].names();
+  for (const auto & name : var_names)
+    if (!dynamic_cast<MooseLinearVariableFVReal *>(_vars[0].getVariable(name)))
+      mooseError("You are trying to add a nonlinear variable to a linear system! The variable "
+                 "which is assigned to the wrong system: ",
+                 name);
+}
+
+void
+LinearSystem::computeLinearSystemTags(const std::set<TagID> & vector_tags,
+                                      const std::set<TagID> & matrix_tags)
+{
+  parallel_object_only();
+
+  TIME_SECTION("LinearSystem::computeLinearSystemTags", 5);
+
+  _fe_problem.setCurrentLinearSystem(_fe_problem.linearSysNum(name()));
+
+  FloatingPointExceptionGuard fpe_guard(_app);
+
+  try
+  {
+    computeLinearSystemInternal(vector_tags, matrix_tags);
+  }
+  catch (MooseException & e)
+  {
+    // The buck stops here, we have already handled the exception by
+    // calling stopSolve(), it is now up to PETSc to return a
```

Comment:
did we call stopSolve in the libmesh solve call?

Reply:
Hmm this does raise a good point. `PARALLEL_CATCH` calls to `FEProblemBase::checkExceptionAndStopSolve` which does `_current_nl_sys->stopSolve(_current_execute_on_flag);`. Perhaps we could have a `_current_solver_system` member and have `stopSolve` be pure virtual in `SolverSystem`. If we don't have a good means to do `stopSolve` for the linear system yet, then we could just do a hard `mooseError` for the moment in the `LinearSystem` implementation of that pure virtual

Reply:
Okay, I added `stopSolve` for the linear system.